clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/BootstrappingVeniceChangelogConsumerDaVinciRecordTransformerImpl.java

```diff
@@ -147,6 +138,29 @@ public CompletableFuture<Void> start(Set<Integer> partitions) {
       }
       return null;
     }, completableFutureThreadPool);
+
+    /*
+     * Avoid waiting on the CompletableFuture to prevent a circular dependency.
+     * When subscribe is called, DVRT scans the entire storage engine and fills pubSubMessages.
+     * Because pubSubMessages has limited capacity, blocking on the CompletableFuture
+     * prevents the user from calling poll to drain pubSubMessages, so the threads populating pubSubMessages
+     * will wait forever for capacity to become available. This leads to a deadlock.
+     */
+    daVinciClient.subscribe(partitions).whenComplete((result, error) -> {
+      if (error != null) {
+        LOGGER.error("Failed to subscribe to partitions: {} for store: {}", partitions, storeName, error);
+        startFuture.completeExceptionally(new VeniceException(error));
+        return;
+      }
+
+      isCaughtUp.set(true);
+      LOGGER.info(
```

Comment:
Don't we also need to complete the `startFuture` successfully here? I also noticed we don't actually use the future returned by start() in our test. Alternatively why don't we just let user use the start future as a signal for is caught up instead of an API they call? To me the former is more accurate since the future belongs to a specific start call for a `Set<Integer> partitions`.

Reply:
Synced offline about this, the intention is to avoid likelihood of dead lock. Also the future is completed when the first message arrives in the buffer.